### PR TITLE
Anstyle-Roff

### LINKED ISSUE #29: Ansi-roff converter
Per conversation in clap/clap#4443 this issue is for discussion around adding
an ansi-to-roff conversion library.

> In epage/anstyle#4, I contrast different ANSI parsing libraries. For a first
> round, feel free to use what is easiest as long as it as a good license. I'm
> less concerned about clap_mangens build times. We can then iterate on it as
> clap needs to adopt an ANSI parser.

> clap_mangen is in its very early stages; nothing is set in stone.

### PR DESCRIPTION
**Pull request by Calder-Ty**
This is an Initial Stab at an implementation for a anstyle to roff converter.  It uses the `roff` crate as the underlying engine. I went with using an intermediate "RoffStyle" struct, rather than converting directly to the `roff::Roff` struct, as the `roff::Roff` expects that there be text associated with it when you add certain effects (bold, italic and plain). Since `anstyle::Style` does not expect text, the `RoffStyle` can be created from `Style` and then text can be added later, prior to rendering the output.

Concerns:

- Effects: ROFF (the tool) has support for _more_ effects than what are implemented here. This is because `roff` (the crate) has limited support for inline escape sequences. Another issue is that `roff` (the crate) only allows for 1 effect to be applied at a time, via inline escapes. Multiple stylings may be possible via ROFF control requests, but they can have unintended results that are not seen with inline escapes. I'll have to experiment more with those.

Fixes #29

### CODE REVIEW

File: crates/anstyle-roff/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+version = "0.1.0"
+description = "Adapt between anstyle and roff"
+license = "MIT OR Apache-2.0"
+repository = "https://github.com/rust-cli/anstyle.git"
+homepage = "https://github.com/rust-cli/anstyle"
+categories = ["command-line-interface"]
+keywords = ["ansi", "terminal", "roff"]
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (epage) [inline crates/anstyle-roff/Cargo.toml:12]: Please remove

File: crates/anstyle-roff/tests/test_roff.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#[test]
+fn test_rgb_color_output() {
+    let expected = include_str!("./roffs/rgb_color.roff");
+    let style = Style::new()
+        .fg_color(Some(Color::Rgb(RgbColor(255, 0, 170))))
+        .bg_color(Some(Color::Rgb(RgbColor(0, 255, 0))));
+
+    let mut roff_style = anstyle_roff::to_roff(style);
+    roff_style.text("test".to_owned());
+    assert_eq!(roff_style.render(), expected)
```
Comment (epage) [inline crates/anstyle-roff/tests/test_roff.rs:12]: Instead of `include_str!` the expected result, you can use `snapbox::assert_eq_path`.  This will make it easier to update everything as we evolve the code/tests.

File: crates/anstyle-roff/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+///     .fg_color(Some(Color::Rgb(RgbColor(255, 0, 170))));
+///
+/// let roff_style = anstyle_roff::to_roff(style);
+/// let expected = r#".defcolor hex_#ff00aa rgb #ff00aa
+/// .gcolor hex_#ff00aa
+/// .fcolor default
+/// "#;
+/// assert_eq!(roff_style.render(), expected);
+/// ```
+pub fn to_roff(style: anstyle::Style) -> RoffStyle {
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:158]: My assumption had been that we'd be parsing a `&str` that contains ANSI escape codes and converting that to roff.

While `clap` currently deals with types that are similar to `anstyle`, the plan is to update it to always be writing out ANSI escape codes and then reinterpret them as needed.

Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:158]: I was patterning this off of `anstyle-yansi` and `anstyl-owo-colors`, with the idea of separating out the stream parsing code into a PR for #4. But now, if I'm understanding correctly, you see this as different than those two crates. `anstyle-yansi`/`owo-colors` are there to provide conversion to other formats, but not necessarily create output, which is what we want from `roff`. That makes sense now that I've thought about it, but want to make sure I'm understanding correctly.

Comment (epage) [reply crates/anstyle-roff/src/lib.rs:158]: Yes, those crates are for integration.  For example, you want `anstyle` in your public API but you are using `yansi` for formatting your strings.

What we are needing for clap is the direct conversion of ansi escape codes to roff.

And I forgot earlier, clap already has [ansi code generation](https://docs.rs/clap/latest/clap/builder/struct.StyledStr.html#method.ansi)

File: crates/anstyle-roff/Cargo.toml (hunk 1 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/Cargo.toml]: Done

File: crates/anstyle-roff/tests/test_roff.rs (hunk 2 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/tests/test_roff.rs]: Done

File: crates/anstyle-roff/src/lib.rs (hunk 3 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs]: I've pushed a new version of this that privately implements a stream parser and uses it. I thought it would be best to put it in the Roff Stream crate for now, while we test it out and then if you like it, I can extract it into the anstyle crate itself for use. It is fairly simple for now.

**Comment by Calder-Ty** [comment]
Thanks for your patience on this.

**Comment by Calder-Ty** [comment]
Any thoughts on these updates?

File: crates/anstyle-roff/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+pub fn to_roff(styled_text: &str) -> Roff {
+    let stream = style_stream::StyledStream::new(styled_text);
+
+    let mut roff_docs = vec![];
+    for style_str in stream {
+        let style = style_str.style;
+        let fg = ansi_color_to_roff(style.get_fg_color());
+        let bg = ansi_color_to_roff(style.get_bg_color());
+        let effect = style.get_effects();
+        let mut roff_style = RoffStyle::new();
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:105]: Do we still need `RoffStyle`?  Seems like it could be a function.

File: crates/anstyle-roff/src/roff_styles.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+impl ControlRequests {
+    /// Control to Create a Color definition
+    pub const CREATE_COLOR: &'static str = "defcolor";
+    /// Roff control request to set background color (fill color)
+    pub const BACKGROUND: &'static str = "fcolor";
+    /// Roff control request to set foreground color (glyph color)
+    pub const FOREGROUND: &'static str = "gcolor";
+}
+
```
Comment (epage) [inline crates/anstyle-roff/src/roff_styles.rs:21]: Seems odd to make these associated constants rather than just putting them in the above mod or in an inner mod

**Review by epage — COMMENTED** [review]
Sorry for the delay, been putting in extra focus to get the `toml` release out

My main concern is that this overall feels like it has a lot of layers that make it harder to follow that seem like they could be removed..  I called out a few in my comments.

**Comment by Calder-Ty** [comment]
>Sorry for the delay, been putting in extra focus to get the toml release out

No problem, sorry for the late response, end of month necessities have got my time limited. I'm going to try to get an update this week for this.

> My main concern is that this overall feels like it has a lot of layers that make it harder to follow that seem like they could be 
> removed.. I called out a few in my comments.

I think those are valid critiques. I'll look at cleaning it up and simplifying it.

File: crates/anstyle-roff/src/style_stream.rs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        Some(Color::BrightMagenta) => Some(AColor::Ansi(AnsiColor::BrightMagenta)),
+        Some(Color::BrightCyan) => Some(AColor::Ansi(AnsiColor::BrightCyan)),
+        Some(Color::BrightWhite) => Some(AColor::Ansi(AnsiColor::BrightWhite)),
+        None => None,
+    }
+}
+
+impl<'text> StyledStream<'text> {
+    pub(crate) fn new(s: &'text str) -> Self {
+        let categorized = cansi::v3::categorise_text(s);
```
[the remark this answers is not in the document]
Comment (Calder-Ty) [reply crates/anstyle-roff/src/style_stream.rs]: My thinking was to isolate the use of cansi to one module in case we want to change the provider for categorizing text. Not using a whole other struct is feasable though. I wrote a func to just return an iterator of `StyledStr`.

File: crates/anstyle-roff/src/roff_styles.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
+#[derive(Debug, Clone)]
+pub(crate) enum Color {
+    Rgb(RgbColor),
+    AnsiColor(String),
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(crate) struct RgbColor(pub u8, pub u8, pub u8);
```
[the remark this answers is not in the document]
Comment (Calder-Ty) [reply crates/anstyle-roff/src/roff_styles.rs]: Eliminated

File: crates/anstyle-roff/src/lib.rs (hunk 4 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs]: Done

File: crates/anstyle-roff/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    let style = styled.style;
+    let mut doc = Roff::new();
+    doc.extend([
+        set_color((&style.get_fg_color(), &style.get_bg_color())),
+        set_effects(styled),
+    ]);
+    doc
+}
+
+fn set_effects(styled: &StyledStr) -> Roff {
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:46]: Ditto about `&mut Roff`

File: crates/anstyle-roff/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        doc.text(vec![italic(styled.text)]);
+        return doc;
+    }
+
+    if effects.is_plain() {
+        doc.text(vec![roff::roman(styled.text)]);
+        return doc;
+    }
+
+    doc
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:72]: It looks like this is dropping unsupported text, which seems bad.

File: crates/anstyle-roff/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    // perhaps by spinning off some of this code
+    let effects = styled.style.get_effects();
+    let mut doc = Roff::new();
+
+    if effects.contains(anstyle::Effects::BOLD) {
+        doc.text(vec![bold(styled.text)]);
+        return doc;
+    }
+
+    if effects.contains(anstyle::Effects::ITALIC) {
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:62]: nit: I prefer making mutually exclusive paths clearer with `else if` rather than relying on fall-through

File: crates/anstyle-roff/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                control_requests::CREATE_COLOR,
+                vec![name.as_str(), "rgb", as_hex(c).as_str()],
+            )
+            .control(control_request, vec![name.as_str()]);
+        }
+
+        Some(Color::Ansi(c)) => {
+            doc.control(control_request, vec![ansi_color_to_roff(c)]);
+        }
+        _ => {
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:99]: I prefer cases to be explicit in matches, where possible

Comment (epage) [inline crates/anstyle-roff/src/lib.rs:99]: `anstyle` can convert some `XTerm` into `Ansi` and then `anstyle-lossy` can convert the rest.

File: crates/anstyle-roff/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    match color {
+        anstyle::AnsiColor::Black => "black",
+        anstyle::AnsiColor::Red => "red",
+        anstyle::AnsiColor::Green => "green",
+        anstyle::AnsiColor::Yellow => "yellow",
+        anstyle::AnsiColor::Blue => "blue",
+        anstyle::AnsiColor::Magenta => "magenta",
+        anstyle::AnsiColor::Cyan => "cyan",
+        anstyle::AnsiColor::White => "white",
+        _ => "default",
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:125]: I feel like `Bright` should either be converted into the color or converted into bold+color, see #12

**Review by epage — COMMENTED** [review]
This is looking great, thanks!

File: crates/anstyle-roff/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    for styled in styled_str::styled_stream(styled_text) {
+        roff_docs.push(as_roff(&styled))
+    }
+
+    let mut doc = Roff::new();
+    doc.extend(roff_docs);
+    doc
+}
+
+fn as_roff(styled: &StyledStr) -> Roff {
```
[the remark this answers is not in the document]
Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:36]: Good catch. After Some refactoring i was able to just inline this function.

File: crates/anstyle-roff/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    let style = styled.style;
+    let mut doc = Roff::new();
+    doc.extend([
+        set_color((&style.get_fg_color(), &style.get_bg_color())),
+        set_effects(styled),
+    ]);
+    doc
```
[the remark this answers is not in the document]
Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:43]: Done

File: crates/anstyle-roff/src/lib.rs (hunk 10 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:62]: Done

File: crates/anstyle-roff/src/lib.rs (hunk 9 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:72]: Hmm. yeah I don't know why i didn't go with a plain else here.

File: crates/anstyle-roff/src/lib.rs (hunk 12 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:125]: Done. Bright now maps to the colors, and there is extra logic when applying the Bold Effect to check if the Color is a Bright variant.

File: crates/anstyle-roff/tests/test_roff.rs [hunk 15]
```diff
@@ -0,0 +1,24 @@
+use colored::Colorize;
```
[the remark this answers is not in the document]
Comment (Calder-Ty) [reply crates/anstyle-roff/tests/test_roff.rs:1]: Doh. I feel so dumb. Not sure why I didn't just use anstyle!

**Comment by Calder-Ty** [comment]
Fixed Commit Lint,
Removed owo-colors dependency
Changed Snapbox dependency to lower version for MSRV check

File: crates/anstyle-roff/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        Some(Color::Ansi(c)) => {
+            doc.control(control_request, vec![ansi_color_to_roff(c)]);
+        }
+        Some(Color::XTerm(c)) => {
+            // Adding Support for XTerm colors, however cansi does not support
+            // XTerm Colors, so this is not executed. If we switch to a provider
+            // That has Xterm support we will also get it for Roff
+            let ansi_color = anstyle_lossy::xterm_to_ansi(*c, Palette::default());
```
Comment (epage) [inline crates/anstyle-roff/src/lib.rs:121]: xterm is losslessly convertable to `Rgb | Ansi`.  Ideally, we'd do that.  Maybe some API changes are needed to make that easier.

Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:121]: I used `anstyle_lossy::xterm_to_ansi` because It looked like it has the same functionality as the `XTermColor::into_ansi` function, but with the added benefit that additional values would get mapped to something _close_ to the requested color. Is there something I'm misunderstanding?

Comment (epage) [reply crates/anstyle-roff/src/lib.rs:121]: Thats a lossy conversion which you don't need but can instead fo:

```rust
(*c).into_ansi().unwrap_or_else(|| anstyle_lossy::xterm_to_rgb(*c, Palette::default());
```
While this accepts a palette, it won't be used and  no lossy conversions are needed.  I'm mixed about adding a `XTerm::into_rgb` as the lookup tables are big but making this more clear its lossless (and not needing a lossy crarte for it) would be nice.

Comment (Calder-Ty) [reply crates/anstyle-roff/src/lib.rs:121]: Ahh, I think I see now. Rather than go straight to Ansi Colors (which would be lossy), use the roff RGB support for XTerm colors that don't map to Ansi, thereby getting all xterm colors supported.

File: crates/anstyle-roff/Cargo.toml [hunk 17]
```diff
@@ -26,3 +26,4 @@ cansi = "2.2.1"
 
 [dev-dependencies]
 snapbox = "0.3.3"
+snapbox-macros = "0.3.0"
```
Comment (epage) [inline crates/anstyle-roff/Cargo.toml:29]: Why was this needed?

**Review by epage — APPROVED** [review]
This is great, thanks!

Just looking for clarification on snapbox and we can merge / release this!

File: crates/anstyle-roff/Cargo.toml (hunk 17 above)
Comment (Calder-Ty) [reply crates/anstyle-roff/Cargo.toml:29]: The CI was failing because of snapbox-macros version requiring a higher MSRV than the CI was allowing.  Hoping that pinning it would fix the issue.

File: crates/anstyle-roff/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,29 @@
+[package]
+name = "anstyle-roff"
+version = "0.1.0"
+description = "Adapt between anstyle and roff"
+license = "MIT OR Apache-2.0"
+repository = "https://github.com/rust-cli/anstyle.git"
+homepage = "https://github.com/rust-cli/anstyle"
+categories = ["command-line-interface"]
+keywords = ["ansi", "terminal", "roff"]
+edition = "2021"
```
Comment (epage) [inline crates/anstyle-roff/Cargo.toml:10]: When #32 lands, can you update this to use workspace inheritance?

Comment (Calder-Ty) [reply crates/anstyle-roff/Cargo.toml:10]: yes I'll do that.

Comment (epage) [reply crates/anstyle-roff/Cargo.toml:10]: Merged

File: Cargo.toml [hunk 19]
```diff
@@ -1,3 +1,11 @@
 [workspace]
 members = ["crates/*"]
 resolver = "2"
+
+[workspace.package]
```
Comment (epage) [inline Cargo.toml:5]: Something about how this commit was made is causing a merge conflict

File: Cargo.toml [hunk 20]
```diff
@@ -1,3 +1,11 @@
...
+license = "MIT OR Apache-2.0"
+repository = "https://github.com/rust-cli/anstyle.git"
+homepage = "https://github.com/rust-cli/anstyle"
+categories = ["command-line-interface"]
+keywords = ["ansi", "terminal"]
```
Comment (epage) [inline Cargo.toml:10]: I intentionally kept these out as we are likely to have deviations on these from crate to crate and want to re-evaluated for each

File: Cargo.toml (hunk 19 above)
Comment (Calder-Ty) [reply Cargo.toml]: Forgot to pull the upstream changes

File: Cargo.toml (hunk 20 above)
Comment (Calder-Ty) [reply Cargo.toml]: Yeah i'm a dope, forgot to pull your changes. Let me get those fixed.

**Comment by epage** [comment]
> It doesn't like the merge commit. Would you like me to rebase and reword the merge commit? I can fix the Clippy mistakes, but not sure how you want to deal with the lint commit failure.

I disallow merge commits, preferring rebasing.

However, I think I'm going to squash this PR so it doesn't matter.

**Comment by Calder-Ty** [comment]
Made the suggested clippy fixes. Thought about trying to fix the merge stuff, but since you are going to squash anyway, i'll let that deal with it.

**Comment by epage** [comment]
Well, I was going to squash...

**Comment by Calder-Ty** [comment]
Were you not able to squash? Hope I didn't screw it up. 

Thanks for the feedback on the PR, been a lot of fun working on it!

**Comment by epage** [comment]
Out of habit I merged, rather than squashed (the button is so easy; its right there).

Nice thing about a low visibility repo in its early stages is I can enable force pushes and rewrite history that only existed for 2 minutes...

**Comment by epage** [comment]
anstyle-roff-v0.1.0 is now out

**Comment by Calder-Ty** [comment]
Awesome!!
